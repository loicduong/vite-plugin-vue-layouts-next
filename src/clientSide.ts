import { posix } from 'node:path'
import { addIndentation } from './utils.js'
import { returnLayoutComponent, returnLayoutRoute } from './layoutReturn.js'

function normalizePath(path: string) {
  path = path.startsWith('/') ? path : `/${path}`
  return posix.normalize(path)
}

interface VirtualModuleCodeOptions {
  layoutDir: string
  defaultLayout: string
  importMode: 'sync' | 'async'
  wrapComponent: boolean
}

async function createVirtualGlob(
  target: string,
  isSync: boolean,
) {
  const g = `"${target}/**/*.vue"`
  return `import.meta.glob(${g}, { eager: ${isSync} })`
}

export async function createVirtualModuleCode(
  options: VirtualModuleCodeOptions,
) {
  const { layoutDir, defaultLayout, importMode } = options

  const normalizedTarget = normalizePath(layoutDir)

  const isSync = importMode === 'sync'

  return /* js */`
  export const createGetRoutes = (router, withLayout = false) => {
      const routes = router.getRoutes()
      if (withLayout) {
          return routes
      }
      return () => routes.filter(route => !route.meta.isLayout)
  }
  
  export const setupLayouts = routes => {
      const layouts = {}
  
      const modules = ${await createVirtualGlob(
    normalizedTarget,
    isSync,
  )}
    
      Object.entries(modules).forEach(([name, module]) => {
          let key = name.replace("${normalizedTarget}/", '').replace('.vue', '')
          layouts[key] = ${isSync ? 'module.default' : 'module'}
      })
      
    function deepSetupLayout(routes, top = true) {
      return routes.map(route => {
        if (route.children?.length > 0) {
          route.children = deepSetupLayout(route.children, false)
        }

        const layout = route.meta?.layout ?? '${options.defaultLayout}'

        const skipLayout = top
          && !route.component
          && route.children?.find(r => (r.path === '' || r.path === '/') && r.meta?.isLayout);
  
        if (skipLayout) {
          return route
        }
            
        if (layout && layouts[layout]) {
          ${addIndentation(options.wrapComponent ? returnLayoutComponent : returnLayoutRoute, 8)}
        }
  
        return route
      })
    }
  
      return deepSetupLayout(routes)
  }`
}
