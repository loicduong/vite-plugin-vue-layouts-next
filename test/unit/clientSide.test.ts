import { describe, expect, it } from 'vitest'
import { createVirtualModuleCode } from '../../src/clientSide.js'

/* describe('virtual module utilities', () => {
}) */

describe('createVirtualModuleCode', () => {
  it('generates code with default layout', async () => {
    const code = await createVirtualModuleCode({
      layoutDir: './layouts',
      defaultLayout: 'default',
      importMode: 'sync',
      // skipTopLevelRouteLayout: false,
      wrapComponent: false,
    })

    expect(code).toMatchSnapshot()
  })

  it('generates code with wrapped components', async () => {
    const code = await createVirtualModuleCode({
      layoutDir: './layouts',
      defaultLayout: 'default',
      importMode: 'async',
      // skipTopLevelRouteLayout: false,
      wrapComponent: true,
    })

    expect(code).toMatchSnapshot()
  })

  it('generates code with top level route layout skipping', async () => {
    const code = await createVirtualModuleCode({
      layoutDir: './layouts',
      defaultLayout: 'default',
      importMode: 'sync',
      // skipTopLevelRouteLayout: true,
      wrapComponent: false,
    })

    expect(code).toMatchSnapshot()
  })
})
