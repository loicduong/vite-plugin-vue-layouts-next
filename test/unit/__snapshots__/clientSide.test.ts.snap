// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createVirtualModuleCode > generates code with default layout 1`] = `
"
  export const createGetRoutes = (router, withLayout = false) => {
      const routes = router.getRoutes()
      if (withLayout) {
          return routes
      }
      return () => routes.filter(route => !route.meta.isLayout)
  }
  
  export const setupLayouts = routes => {
      const layouts = {}
  
      const modules = import.meta.glob("/layouts/**/*.vue", { eager: true })
    
      Object.entries(modules).forEach(([name, module]) => {
          let key = name.replace("/layouts/", '').replace('.vue', '')
          layouts[key] = module.default
      })
      
    function deepSetupLayout(routes, top = true) {
      return routes.map(route => {
        if (route.children?.length > 0) {
          route.children = deepSetupLayout(route.children, false)
        }

        const layout = route.meta?.layout ?? 'default'

        const skipLayout = top
          && !route.component
          && route.children?.find(r => (r.path === '' || r.path === '/') && r.meta?.isLayout);
  
        if (skipLayout) {
          return route
        }
            
        if (layout && layouts[layout]) {
          
/** @type {import('vue-router').RouteRecordRaw} */
const layoutRoute = {
        path: route.path,
        component: layouts[layout],
        meta: { ...route.meta, isLayout: true },
        // Handle root path specially to avoid infinite nesting
        children: top && route.path === '/' 
        ? [route] 
        : [{ ...route, path: '', meta: { ...route.meta, isLayout: false } }]
}
return layoutRoute

        }
  
        return route
      })
    }
  
      return deepSetupLayout(routes)
  }"
`;

exports[`createVirtualModuleCode > generates code with top level route layout skipping 1`] = `
"
  export const createGetRoutes = (router, withLayout = false) => {
      const routes = router.getRoutes()
      if (withLayout) {
          return routes
      }
      return () => routes.filter(route => !route.meta.isLayout)
  }
  
  export const setupLayouts = routes => {
      const layouts = {}
  
      const modules = import.meta.glob("/layouts/**/*.vue", { eager: true })
    
      Object.entries(modules).forEach(([name, module]) => {
          let key = name.replace("/layouts/", '').replace('.vue', '')
          layouts[key] = module.default
      })
      
    function deepSetupLayout(routes, top = true) {
      return routes.map(route => {
        if (route.children?.length > 0) {
          route.children = deepSetupLayout(route.children, false)
        }

        const layout = route.meta?.layout ?? 'default'

        const skipLayout = top
          && !route.component
          && route.children?.find(r => (r.path === '' || r.path === '/') && r.meta?.isLayout);
  
        if (skipLayout) {
          return route
        }
            
        if (layout && layouts[layout]) {
          
/** @type {import('vue-router').RouteRecordRaw} */
const layoutRoute = {
        path: route.path,
        component: layouts[layout],
        meta: { ...route.meta, isLayout: true },
        // Handle root path specially to avoid infinite nesting
        children: top && route.path === '/' 
        ? [route] 
        : [{ ...route, path: '', meta: { ...route.meta, isLayout: false } }]
}
return layoutRoute

        }
  
        return route
      })
    }
  
      return deepSetupLayout(routes)
  }"
`;

exports[`createVirtualModuleCode > generates code with wrapped components 1`] = `
"
  export const createGetRoutes = (router, withLayout = false) => {
      const routes = router.getRoutes()
      if (withLayout) {
          return routes
      }
      return () => routes.filter(route => !route.meta.isLayout)
  }
  
  export const setupLayouts = routes => {
      const layouts = {}
  
      const modules = import.meta.glob("/layouts/**/*.vue", { eager: false })
    
      Object.entries(modules).forEach(([name, module]) => {
          let key = name.replace("/layouts/", '').replace('.vue', '')
          layouts[key] = module
      })
      
    function deepSetupLayout(routes, top = true) {
      return routes.map(route => {
        if (route.children?.length > 0) {
          route.children = deepSetupLayout(route.children, false)
        }

        const layout = route.meta?.layout ?? 'default'

        const skipLayout = top
          && !route.component
          && route.children?.find(r => (r.path === '' || r.path === '/') && r.meta?.isLayout);
  
        if (skipLayout) {
          return route
        }
            
        if (layout && layouts[layout]) {
          
if (!route.component) {
        return route
}
/** @type {import('vue-router').RouteRecordRaw} */
const wrappedRoute = {
        ...route,
        component: h('div', [
        h(layouts[layout].layout),
        h(layouts[layout].isSync ? defineComponent(() => route.component()) : defineAsyncComponent(() => route.component())),
        ]),
        meta: {
        ...route.meta,
        isLayout: true
        }
}
return wrappedRoute

        }
  
        return route
      })
    }
  
      return deepSetupLayout(routes)
  }"
`;
